File: apps/cli/src/emoji.zig
const std = @import("std");
const utils = @import("utils.zig");

const CACHE_FILE = ".genmoji/gitmoji.json";

const Gitmoji = struct {
    emoji: []u8,
    code: []u8,
    description: []u8,
    name: []u8,
};

const Response = struct {
    gitmojis: []Gitmoji,
};

fn getPath(allocator: std.mem.Allocator) !?[]u8 {
    var env = try std.process.getEnvMap(allocator);
    defer env.deinit();

    const homedir = env.get("HOME") orelse {
        return null;
    };

    return try std.fmt.allocPrint(allocator, "{s}/{s}", .{ homedir, CACHE_FILE });
}

fn checkIfExists(path: []u8) bool {
    std.fs.accessAbsolute(path, .{}) catch {
        return false;
    };

    return true;
}

fn fetchFromCache(allocator: std.mem.Allocator) !?Response {
    const path = try getPath(allocator) orelse {
        return null;
    };

    if (!checkIfExists(path)) {
        return null;
    }

    const file = try std.fs.openFileAbsolute(path, .{});
    defer file.close();

    const stat = try file.stat();
    const content = try file.readToEndAlloc(allocator, stat.size);

    return try std.json.parseFromSliceLeaky(Response, allocator, content, .{ .allocate = .alloc_always, .ignore_unknown_fields = true });
}

fn writeToCache(allocator: std.mem.Allocator, data: Response) !void {
    const path = try getPath(allocator) orelse {
        return;
    };

    try std.fs.makeDirAbsolute(std.fs.path.dirname(path).?);
    const file = try std.fs.createFileAbsolute(path, .{ .truncate = true });
    defer file.close();

    const string = try std.json.stringifyAlloc(allocator, data, .{});
    try file.writeAll(string);
}

fn fetchFromRemote(allocator: std.mem.Allocator) !Response {
    const response = try utils.fetch(Response, "https://gitmoji.dev/api/gitmojis", .{
        .allocator = allocator,
        .method = .GET,
    });

    writeToCache(allocator, response) catch |err| {
        std.log.debug("Warning: Ignoring error while writing to cache: {}\n", .{err});
    };

    return response;
}

pub fn fetchGitmojis(allocator: std.mem.Allocator) !Response {
    const cache = fetchFromCache(allocator) catch |err| {
        std.log.debug("Warning: Ignoring error while fetching from cache: {}\n", .{err});

        return try fetchFromRemote(allocator);
    };

    if (cache) |response| {
        return response;
    }

    return try fetchFromRemote(allocator);
}

File: apps/cli/emoji.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"os"
)

type Gitmoji struct {
	Emoji       string `json:"emoji"`
	Code        string `json:"code"`
	Description string `json:"description"`
	Name        string `json:"name"`
}

type Response struct {
	Gitmojis []Gitmoji `json:"gitmojis"`
}

func isCached(path string) bool {
	if _, err := os.Stat(path); os.IsNotExist(err) {
		return false
	}

	return true
}

func fetchFromCache(path string) ([]Gitmoji, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var response Response
	err = json.NewDecoder(file).Decode(&response)
	if err != nil {
		return nil, err
	}

	return response.Gitmojis, nil
}

func writeToCache(path string, response Response) error {
	file, err := os.Create(path)
	if err != nil {
		return err
	}
	defer file.Close()

	err = json.NewEncoder(file).Encode(&response)
	if err != nil {
		return err
	}

	return nil
}

func fetchGitmojis() ([]Gitmoji, error) {
	dirname, err := os.UserHomeDir()
	if err != nil {
		log.Fatal(err)
	}

	var path string = dirname + "/.genmoji/gitmojis.json"
	if isCached(path) {
		return fetchFromCache(path)
	}

	res, err := http.Get("https://gitmoji.dev/api/gitmojis")
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()

	var response Response
	if err := json.NewDecoder(res.Body).Decode(&response); err != nil {
		return nil, err
	}

	if err := writeToCache(path, response); err != nil {
		return nil, err
	}

	return response.Gitmojis, nil
}
